### PR TITLE
Use generic collections in System.ComponentModel.TypeConverter implementation

### LINKED ISSUE #17217: Use generic collections in System.ComponentModel.TypeConverter implementation instead of non-generic

### PR DESCRIPTION
**Pull request by MattKotsenas**
Fix #17217 by replacing non-generic collections classes with their generic counterparts. Of specific interest are:

- The EditorTable related collections are public, and thus can't be converted without changing the public API
- In a couple of places we have public APIs that return the Hashtable's enumerator, so we need to leave those as-is so the enumerated items are `DictionaryEntry` and not `KeyValuePair`
- In places where the `Hashtables` were static, I verified that every use was either already under a lock, or switched to a `ConcurrentDictionary` to more closely match the Hashtable semantics of allowing multiple readers and a single writer

### CODE REVIEW

**Comment by MattKotsenas** [comment]
Of particular note is that I don't know the ramifications of the LicenseManager change. If the serialized value is only guaranteed for a single .NET version, then I assume this is safe? If these are meant to be persisted across runs however then I'm happy to back the change out.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: One reason we haven't mass replaced `Hashtable` with `Dictionary<K,V>` is because the thread-safety differences between the two implementations.

See https://stackoverflow.com/questions/10255456/difference-between-dictionary-and-hashtable-in-terms-of-thread-safety

Comment (MattKotsenas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: Good point! I recently learned about the difference in thread safety for the two types, but hadn't considered the implications until you mentioned it.

I'll take another look through all the changes with this in mind. Are there any specific guidelines you'd like me to follow? For instance, do we want to aim to remove all the non-generic uses (like the issue says), or instead is the issue overly optimistic, and in multi-threaded scenarios we should instead leave the Hashtables as-is and avoid the churn?

Your thoughts are appreciated. Thanks!

Comment (teo-tsirpanis) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: Should it be a `ConditionalWeakTable` here as well? Storing `Type`s in a static doctionary would inhibit unloadability. And CWT is thread-safe.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs
Comment (teo-tsirpanis) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs:24]: Here as well.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs
Comment (teo-tsirpanis) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:59]: And here.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs
Comment (teo-tsirpanis) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:31]: And here.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs
Comment (stephentoub) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: > in multi-threaded scenarios we should instead leave the Hashtables as-is and avoid the churn?

If the Hashtable is ever read outside of a lock but updated inside of a lock, it should remain a Hashtable or a ConcurrentDictionary.

Comment (stephentoub) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: Another difference to keep in mind is if anything from the collection is ever exposed, e.g. an enumerator.  We've had accidental breaking changes in the past where a Hashtable was changed to a Dictionary and that in turn changed the type of the values yielded from an enumerator.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerCollection.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerCollection.cs:43]: This is a behavioral change and an increase in cost.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerOptionService.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerOptionService.cs:217]: Another example of increasing cost.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:30]: WeakHashtable has different semantics than ConditionalWeakTable, in particular around when things are dropped from the collection.  ConditionalWeakTable uses weak references for all the keys, which means something added could be removed on the very next GC; in contrast, WeakHashtable had its own policy for when keys would be scavenged based on memory thresholds.  I don't think we can just change one to the other without a comprehensive performance analysis.

**Comment by stephentoub** [comment]
Thanks for working on this.  I suggest scaling back the PR to just the replacements that are trivial to prove are correct.  There are a myriad of functional and performance differences between what's being replaced and the replacements, and I'm concerned we'll find a number of regressions that slip through as a result.

**Comment by MattKotsenas** [comment]
Thanks everyone for the feedback! I've pushed a less ambitious version that converts just two cases:

- Private instance members that don't (at least appear to) have any multi-threading
- Private statics where based on usage a `ConcurrentDictionary` looks appropriate

Additional feedback and thoughts are very welcome. My goal is less change for change's sake, and more to address the spirit of the linked issue (and hopefully close it out). If anyone has any thoughts, questions, or concerns, please let me know. Thanks!

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs
Comment (MattKotsenas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: Thanks for the info! I've gone back with an eye towards concurrent use, and also did find a case where we expose an enumerator, so I've reverted those changes as well.

Comment (MattKotsenas) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: I defer to @stephentoub for thoughts here. I was trying to avoid intentionally introducing behavior changes in this PR, so I wasn't intending to make a change like that.

**Review by teo-tsirpanis — CHANGES_REQUESTED** [review]
Static dictionaries that contain `Type` objects must be `ConditionalWeakTable`s, not (`Concurrent`)`Dictionary`ies. Otherwise the types are rooted which inhibits unloadability and leads to bugs like #62050.

I have marked some places that should be changed, could you please do that? You can take advantage of the `GetValue` method; it works similarly to `ConcurrentDictionary.GetOrAdd`.

`s_defaultAttributes` is safe to change to CWT as well.

_Edit_: Oh, you had responded to a comment of mine. I thought you hadn't seen them.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs
Comment (teo-tsirpanis) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: They could go in a separate PR but you are already changing this line, better do it the right way immediately. But let's wait for stephentoub.

**Comment by stephentoub** [comment]
> Static dictionaries that contain Type objects must be ConditionalWeakTables, not (Concurrent)Dictionaryies. Otherwise the types are rooted which inhibits unloadability and leads to bugs like #62050.

As long as this PR isn't adding additional rooting, I would prefer any changes to address concerns around unloadability be done separately in a dedicated PR.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs
Comment (stephentoub) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:23]: Addressing unrelated, untested concerns around unloadability is orthogonal to this PR.  Let's not conflate the issues.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:30]: I'm not sure I can trivially prove the changes in this file are correct. Can we revert this file's changes?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:29]: I'm not sure I can trivially prove the changes in this file are correct. Can we revert this file's changes?

**Review by eerhardt — COMMENTED** [review]
Thanks for the PR @MattKotsenas.

Most of the files look fine. But 2 of them I'm not 100% sure are better. The 2 I tagged already have locking around modification, which we aren't changing the locking in this PR. I'm not sure we can trivially prove that changing from `Hashtable` to `ConcurrentDictionary` is beneficial. Can we revert those two files? We aren't actively investing in System.ComponentModel, and changes here would introduce more risk than the value these changes bring.

Other than that I think the rest of the changes are OK.

**Comment by eerhardt** [comment]
@MattKotsenas - any response to my above feedback? If you aren't actively working on this PR, I can close it until it is active again.

**Comment by MattKotsenas** [comment]
Hi @eerhardt ! Yes, I'll back out those changes as well and update to resolve conflicts today or tomorrow. Thanks again for all the feedback!

**Comment by MattKotsenas** [comment]
@eerhardt updated! Please take a look when you get a chance and let me know if you have any thoughts / suggestions. Thanks!

**Review by eerhardt — APPROVED** [review]
I think this slimmed down change is now fine. And with it we should be able to close the long-standing issue.

Thanks for the contribution, @MattKotsenas!

@stephentoub - I will wait for your approval before removing the "No Merge" label. Let me know what you think.

**Review by stephentoub — APPROVED** [review]
Thanks. Seems ok.  All of those GetValueOrDefault calls are extension method calls resulting in virtual dispatch via IReadOnlyDictionary; it's probably ok given the library.

**Comment by eerhardt** [comment]
Test failure is https://github.com/dotnet/runtime/issues/64389. Merging.

**Pull request merged**
